packages/loopring_v3/contracts/amm/libamm/AmmTransactionReceiver.sol

```diff
@@ -94,19 +101,110 @@ library AmmTransactionReceiver
             signature.offset := add(signature.offset, 0x20)
         }
         if (txType == AmmData.PoolTxType.JOIN) {
+            ctx.approveAmmUpdates(true);
             S.processJoin(
                 ctx,
                 abi.decode(data, (AmmData.PoolJoin)),
                 signature
             );
+            ctx.approveAmmUpdates(false);
         } else if (txType == AmmData.PoolTxType.EXIT) {
+            ctx.approveAmmUpdates(true);
             S.processExit(
                 ctx,
                 abi.decode(data, (AmmData.PoolExit)),
                 signature
             );
+            ctx.approveAmmUpdates(false);
+        } else if (txType == AmmData.PoolTxType.SET_VIRTUAL_BALANCES) {
+            ctx.approveAmmUpdates(true);
+            ctx.vTokenBalancesL2 = ctx.settings.controller.getVirtualBalances(ctx.tokenBalancesL2, ctx.vTokenBalancesL2);
+            ctx.approveAmmUpdates(false);
+        } else if (txType == AmmData.PoolTxType.DEPOSIT) {
+            AmmData.PoolDeposit memory poolDeposit = abi.decode(data, (AmmData.PoolDeposit));
+            require(poolDeposit.amounts.length == ctx.tokens.length, "INVALID_DEPOSIT_AMOUNTS");
+            for (uint i = 0; i < ctx.tokens.length; i++) {
+                deposit(S, ctx.tokens[i].addr, poolDeposit.amounts[i]);
+            }
+        } else if (txType == AmmData.PoolTxType.WITHDRAW) {
+            AmmData.PoolWithdrawal memory poolWithdrawal = abi.decode(data, (AmmData.PoolWithdrawal));
```

Comment:
Where is such a AmmData.PoolTxType.WITHDRAW authorized? For Joins/exits, we do L2 signature validation, for this  WITHDRAW, do we also validate a L2 signature? It seems we don't have a L2 EDDSA key for the pool contract.

Reply:
The only authorization needed is passing the checks in this contract, no L1 or L2 signatures are used. Only the operator can deposit/withdraw to/from the pool usable by the asset manager so no other authorization is required. Remember that transactions that are processed by the pool contract are automatically approved on L1 (and these are all conditional transactions which don't need any L2 signatures).